File: Javascript/linked-list.js
function LinkedList() {
    let Node = function (element) {
        this.element = element;
        this.next = null;
    }
    let length = 0;
    let head = null;

    this.append = function (element) {
        let node = new Node(element);
        let current = null;
        if (head === null) {
            head = node;
        }
        else{
            current = head;
            while (current.next) {
                current = current.next;
            }
            current.next = node;
        }
        length++;
    }
    this.insert = function (position, element) {
        if (position >= 0 && position <= length) {
            let node = new Node(element),
            current = head,
            previous,
            index = 0;

            if (position === 0) {
                node.next = current;
                head = node;
            }
            else {
                while (index++ < position) {
                    previous = current;
                    current = current.next;
                }
                node.next = current;
                previous.next = node;
            }
            length++;
            return true;
        }
        else {
            return false;
        }
    }
    this.removeAt = function (position) {
        if (position > -1 && position < length) {
            let current = head,
            previous,
            index = 0;

            if (position === 0) {
                head = current.next;
            }
            else {
                while (index++ < position) {
                    previous = current;
                    current = current.next;
                }
                previous.next = current.next;
            }
            length--;
            return current.element;
        }
        else {
            return null;
        }
    }
    this.remove = function (element) {
        let index = this.indexOf(element);
        this.removeAt(index);
    }
    this.indexOf = function (element) {
        let current = head;
        let index = 0;
        while (current) {
            if (element === current.element) {
                return index;
            }
            index++;
            current = current.next;
        }
        return -1;
    }
    this.isEmpty = function () {
      return length === 0;  
    }
    this.size = function () {
        return length;
    }
    this.getHead = function () {
        return head;
    }
    this.toString = function () {
        let current = head;
        string = '';
        while (current) {
            string += current.element + (current.next ? '->' : ' ');
            current = current.next;
        }
     return string; 
    }
}

let list = new LinkedList();
console.log(list.toString())
list.append(8);
list.append(2);
list.append(3);
list.append(12);
console.log(list.toString());
list.append(6);
list.insert(3, 4);
console.log(list.toString());
list.append(1);
list.insert(0, 9);
console.log(list.toString());
list.removeAt(2);
console.log(list.toString());
list.remove(12);
console.log(list.toString());
list.remove(22);
console.log(list.toString());
list.remove(1);
console.log(list.toString());
list.remove(22);
console.log(list.isEmpty());
console.log(list.size());
console.log(list.getHead());

File: Python/linked-list.py
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None


class LinkedList:
    def __init__(self):
        self.head = None
        self.length = 0

    def add(self, value):
        temp = self.head
        node = Node(value)
        node.next = temp
        self.head = node
        self.length += 1

    def addLast(self, value):
        if self.head == None:
            self.head = Node(value)
        else:
            temp = self.head
            while temp.next:
                temp = temp.next
            temp.next = Node(value)
        self.length += 1


    def deleteFirst(self):
        temp = self.head.next
        self.head = temp
        self.length -= 1

    def delete(self, index):
        prev = None
        node = self.head
        i = 0
        while node and i < index:
            prev = node
            node = node.next
            i += 1
        if index == i:
            self.length -= 1
            if prev == None:
                self.head = node.next
            else:
                prev.next = node.next
        else:
                raise Exception('Index not found')

    def printList(self):
        node = self.head
        items = ''
        while node:
            items += str(node.value) + '->'
            node = node.next 
        print(items)
    
def main():
    list = LinkedList()
    list.add(9)
    list.printList()
    list.add(1)
    list.add(8)
    list.add(5)
    list.printList()
    list.addLast(10)
    list.add(7)
    list.addLast(12)
    list.printList()
    list.deleteFirst()
    list.printList()
    list.deleteFirst()
    list.printList()
    list.delete(2)
    list.add(28)
    list.add(15)
    list.printList()
    list.delete(5)
    list.printList()

if __name__ == "__main__":
    main()

